### PR TITLE
involve rustup when determining toolchain paths

### PR DESCRIPTION
**Pull request by drahnr**
Start of impl to handle #87

### CODE REVIEW

**Comment by drahnr** [comment]
Does not cover the case `cargo +nightly b`

**Comment by drahnr** [comment]
Is the restriction to rsut `1.36` necessary?

**Comment by drahnr** [comment]
use a lookup for the `compiler_path` to a `Box<dyn CompilerProxy>` which is used for the exact compiler lookup depending on the `cwd` current working dir.
`RustupProxy` runs `rustup which rustc` and creates a `Rust` / `Box<dyn Compiler>` instance.
This information is then stored with modifaction time stored as if the Proxy never existed.

Current ugliness:

* compiler type needs to be known so the correct proxy can be picked up - this is all mingled in `detect_compiler` right now, which returns a `Rust` and `RustupProxy` (or more precisely: `(Box<dyn Compiler>, Option<Box<dyn CompilerProxy>>)`

**Comment by drahnr** [comment]
This was more work than anticipated and the futures pre-stabilization caused quite a few not anticipated changes.
I'd like some feedback on the allover idea, and maybe on how to reduce the clutter a bit.

**Comment by drahnr** [comment]
Verified this at least solves #663 .

**Comment by drahnr** [comment]
@froydnj if you could give some feedback so I can move this forward, that'd be great

File: src/server.rs [hunk 1]
```diff
@@ -817,82 +850,175 @@ where
...
+                        f_ok(Ok(info))
+                    }
+                    None => {
+                        trace!("compiler_info cache miss");
+                        // Check the compiler type and return the result when
+                        // finished. This generally involves invoking the compiler,
+                        // so do it asynchronously.
+
+                        // this must be done with the passed `path`, since that is the proxy
+                        // only this will work in all cases
```
Comment (froydnj) [inline src/server.rs]: I'm confused about this comment.  What is it trying to explain?

File: src/server.rs [hunk 2]
```diff
@@ -600,6 +600,26 @@ impl<C: CommandCreatorSync> SccacheServer<C> {
     }
 }
 
+/// entry of the compiler cache
+struct CompilerCacheEntry<C : CommandCreatorSync> {
+    /// compiler argument trait obj
+    pub compiler : Box<dyn Compiler<C>>,
+    /// when did this entry get inserted into the cache
+    pub mtime : FileTime,
```
Comment (froydnj) [inline src/server.rs]: Is this comment correct?  I would expect this field to be related to the modification time of `compiler` in some fashion.

File: src/server.rs [hunk 3]
```diff
@@ -817,82 +850,175 @@ where
...
+                        if let Some(archive) = client.get_custom_toolchain(&resolved_compiler_path) {
+                            match metadata(&archive)
+                                .map(|attr| FileTime::from_last_modification_time(&attr))
+                            {
+                                Ok(mtime) => Some((archive, mtime)),
+                                _ => None,
+                            }
+                        } else {
+                            None
+                        }
```
Comment (froydnj) [inline src/server.rs]: Feels like this could be done more succinctly with some `.map()` calls.

File: .travis.yml [hunk 4]
```diff
@@ -34,7 +34,7 @@ before_script:
 
 script:
   - cargo build --verbose --features="all ${EXTRA_FEATURES}" || exit 1
-  - RUST_BACKTRACE=1 cargo test --all --verbose --no-default-features --features="${EXTRA_FEATURES}" 
+  - RUST_BACKTRACE=1 cargo test --all --verbose --no-default-features --features="${EXTRA_FEATURES}"
```
Comment (froydnj) [inline .travis.yml]: Please separate this out into its own pull request.

File: Cargo.toml [hunk 5]
```diff
@@ -67,7 +67,7 @@ serde_derive = "1.0"
 serde_json = "1.0"
 strip-ansi-escapes = "0.1"
 tar = "0.4"
-tempdir = "0.3.4"
+tempdir = "0.3"
```
Comment (froydnj) [inline Cargo.toml]: Please separate this out into its own pull request.

File: src/compiler/compiler.rs [hunk 6]
```diff
@@ -1065,13 +1160,14 @@ mod test {
         let f = TestFixture::new();
         let creator = new_creator();
         let pool = CpuPool::new(1);
+        let cwd : PathBuf = "/tmp".into();
         next_command(
             &creator,
             Ok(MockChild::new(exit_status(0), "foo\nbar\ngcc", "")),
         );
-        let c = detect_compiler(&creator, &f.bins[0], &[], &pool, None)
+        let c = detect_compiler(creator.clone(), &f.bins[0], f.tempdir.path(), &[], &pool, None)
```
Comment (froydnj) [inline src/compiler/compiler.rs]: Can't you just pass `creator` here and all the places following?

File: src/server.rs [hunk 7]
```diff
@@ -600,6 +600,26 @@ impl<C: CommandCreatorSync> SccacheServer<C> {
...
+struct CompilerCacheEntry<C : CommandCreatorSync> {
+    /// compiler argument trait obj
+    pub compiler : Box<dyn Compiler<C>>,
+    /// when did this entry get inserted into the cache
+    pub mtime : FileTime,
+    /// distributed compilation extra info
+    pub dist_info : Option<(PathBuf, FileTime)>,
+}
+
+impl<C> From<(Box<dyn Compiler<C>>, FileTime, Option<(PathBuf, FileTime)>)> for CompilerCacheEntry<C> where C: CommandCreatorSync {
```
[the remark this answers is not in the document]
Comment (drahnr) [reply src/server.rs]: I don't have a strong opinion about this. I see your point getting rid of this from, it essentially moves `(...).into()` to a `Self::new(...)` call. IIRC, that was the only use case, so changing that should be trivial.

File: src/server.rs [hunk 8]
```diff
@@ -817,82 +850,175 @@ where
...
+                self.creator.clone(),
+                cwd.clone(),
+                env.as_slice(),
+            );
+            Box::new(fut)
+        } else {
+            f_ok(None)
+        };
+
+        drop(compiler_proxies_borrow);
```
[the remark this answers is not in the document]
Comment (drahnr) [reply src/server.rs]: Doable, but this would cause a scope with return value and hamper readability a bit (I had that before).

File: src/server.rs (hunk 3 above)
Comment (drahnr) [reply src/server.rs]: IIRC the `&archive` borrow by `metadata` does not allow returning it by value - avoids a clone :)

File: src/server.rs (hunk 1 above)
Comment (drahnr) [reply src/server.rs]: True, clarification pending..

File: src/server.rs (hunk 2 above)
Comment (drahnr) [reply src/server.rs]: You are correct, the comment is wrong.

File: src/server.rs [hunk 9]
```diff
@@ -817,82 +850,175 @@ where
...
     }
 
+
+
     /// Look up compiler info from the cache for the compiler `path`.
     /// If not cached, determine the compiler type and cache the result.
     fn compiler_info(
         &self,
         path: PathBuf,
+        cwd: PathBuf,
```
[the remark this answers is not in the document]
Comment (drahnr) [reply src/server.rs]: I think you are correct, this can be removed.

File: .travis.yml [hunk 10]
```diff
@@ -7,8 +7,8 @@ matrix:
     - os: osx
 
     # rustc version compat
-    - rust: 1.36.0 # oldest supported version, keep in sync with README.md
-    - rust: 1.36.0
+    - rust: 1.40.0 # oldest supported version, keep in sync with README.md
+    - rust: 1.40.0
```
[the remark this answers is not in the document]
Comment (drahnr) [reply .travis.yml]: but that would add 

```
map(x : Option<Option<_>>) {
match x {
 Some(inner) => inner,
 None => None,
}
}
```
when the code already has a certain complexity

File: src/compiler/compiler.rs [hunk 11]
```diff
@@ -894,28 +922,93 @@ where
...
+    let cwd = cwd.to_owned().clone();
+    Box::new(
+        rustc_vv
+            .and_then(move |rustc_vv| match rustc_vv {
+            Some(Ok(rustc_verbose_version)) => {
+                debug!("Found rustc");
+
+                Box::new(
+                    RustupProxy::find_proxy_executable::<T>("rustup", creator, &env1)
+                        .and_then(move |proxy : Result<RustupProxy>| -> SFuture<(Option<RustupProxy>, PathBuf)> {
```
[the remark this answers is not in the document]
Comment (drahnr) [reply src/compiler/compiler.rs]: How would you determine if `rustc` is a proxy? Currently it's attempted if `rustup` is installed, and if, if try to see if there is an overwrite of the default toolchain and then move from there.

File: src/compiler/compiler.rs (hunk 6 above)
Comment (drahnr) [reply src/compiler/compiler.rs]: Not sure I understand what you mean

File: src/compiler/rust.rs [hunk 12]
```diff
@@ -483,6 +493,99 @@ where
...
+        &self,
+        mut creator: T,
+        _cwd: PathBuf,
+        env: &[(OsString,OsString)],
+    ) -> SFuture<Option<(PathBuf, FileTime)>> {
+
+        let proxy_executable = self.proxy_executable.clone(); // XXX is there a better way to do this?
+
+        let mut child = creator.new_command_sync(&proxy_executable);
+        child.env_clear().envs(ref_env(env)).args(&["which", "rustc"]);
```
[the remark this answers is not in the document]
Comment (drahnr) [reply src/compiler/rust.rs]: Effectively this is done by `env`, so either use cwd to override the `PWD` env var, or pass the env vars as they are. I am not 100% sure this is correct workspaces and nested `Cargo.toml` projects

File: src/server.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (froydnj) [reply src/server.rs]: I just have a small allergy to using tuples (and `.0`, `.1`, etc.) when you could have structures and have real names, so anytime that real structures can be used, I think that's a definite improvement over tuples.

File: src/compiler/compiler.rs (hunk 6 above)
Comment (froydnj) [reply src/compiler/compiler.rs]: No, I mean calling `let c = detect_compiler(creator, ...)...`, since -- having not actually had the borrow checker check this -- I don't think `creator` is actually used after this point.

File: src/server.rs (hunk 3 above)
Comment (froydnj) [reply src/server.rs]: Ah, I missed the `archive` use further down.  OK, I can see how that might get complicated.

File: src/compiler/compiler.rs (hunk 6 above)
Comment (drahnr) [reply src/compiler/compiler.rs]: Understood.

File: src/compiler/compiler.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (froydnj) [reply src/compiler/compiler.rs]: This is what we do in Firefox to figure out whether `rustc` is a proxy: https://searchfox.org/mozilla-central/rev/c79c0d65a183d9d38676855f455a5c6a7f7dadd3/build/moz.configure/rust.configure#23-80

It would be somewhat more complicated than what's already being done, but I think it would be more correctly generally.

[the remark this answers is not in the document]
Comment (drahnr) [reply src/compiler/compiler.rs]: That would add another layer of indirection, and an additional process to spawn, but I agree that this would be the best approach.
Can we merge this and make that check a subsequent PR?

File: src/server.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (drahnr) [reply src/server.rs]: Done

**Comment by drahnr** [comment]
@froydnj I addressed all concerns and verified it works using `spearow/juice` as a demo project - there are still 2 open comments :)

File: src/compiler/compiler.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (luser) [reply src/compiler/compiler.rs]: I suggested a simple test in the issue:
https://github.com/mozilla/sccache/issues/87#issuecomment-587547770

[the remark this answers is not in the document]
Comment (drahnr) [reply src/compiler/compiler.rs]: But it already uses the which crate? That would only resolve if rustup would be right next beside it, but not for the case where rustc is installed via os packages but also a rustc version is installed via rustup. In that case the test as posted in #87 is not yielding viable new information. Hence I would prefer to do this properly as @froydnj suggested, but as a follow up / once everything else in this PR is merge ready.

[the remark this answers is not in the document]
Comment (luser) [reply src/compiler/compiler.rs]: The point is not to detect "is rustup installed at all?" but "is this *particular* rustc we're being invoked with a rustup shim?"

[the remark this answers is not in the document]
Comment (drahnr) [reply src/compiler/compiler.rs]: Ok, so the assumption here is that `rustup` and `rustc` being in the same dir are a sufficient discriminator that the rust is a shim. While this is cheap I would still prefer the behaviour check.

**Comment by drahnr** [comment]
@froydnj rebased

**Comment by drahnr** [comment]
@froydnj can we clarify the last point, (either implement one of both options or merge this as is) unless there are any remaining issues? that'd be much appreciated - thanks!

**Comment by drahnr** [comment]
@froydnj @luser I implemented a hybrid solution of what both you suggested, after all they are not really exclusive and work very well hand in hand. Given the fact that it is only run once per task on the client, the amount of additional work required should be negligible since it's cached for subsequent compiles.

**Comment by drahnr** [comment]
b4aabf0 was added to make `+beta` happy

File: .travis.yml (hunk 10 above)
[the remark this answers is not in the document]
Comment (drahnr) [reply .travis.yml]: For std futures this will need to be bumped anyways, so I don't see this as a problem.

File: src/compiler/compiler.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (drahnr) [reply src/compiler/compiler.rs]: Implemented.

**Comment by drahnr** [comment]
@froydnj  @chmanchester is there anything else I can do to get this merged rather sooner than later?

**Comment by froydnj** [comment]
> @froydnj @chmanchester is there anything else I can do to get this merged rather sooner than later?

Do all my other work so I have a clear time to look at this again?  :)

I will look at it sometime this week.

File: src/server.rs [hunk 13]
```diff
@@ -613,6 +638,11 @@ struct SccacheService<C: CommandCreatorSync> {
     /// A cache of known compiler info.
     compilers: Rc<RefCell<CompilerMap<C>>>,
 
+    /// map the cwd with compiler proxy path to a proxy resolver, which
+    /// will dynamically resolve the input compiler for the current context
+    /// (usually file or current working directory)
```
Comment (froydnj) [inline src/server.rs]: This comment needs to describe what the `FileTime` in the map value represents.

File: src/server.rs [hunk 14]
```diff
@@ -823,68 +855,169 @@ where
...
+                    cwd.clone(),
+                    env.as_slice(),
+                );
+                Box::new(fut)
+            } else {
+                f_ok(None)
+            }
+        };
+
+        // use the supplied compiler path as fallback, lookup it's modification time too
```
Comment (froydnj) [inline src/server.rs]: Nit: "its"

File: lru-disk-cache/src/lib.rs [hunk 15]
```diff
@@ -87,9 +87,8 @@ impl fmt::Display for Error {
 impl StdError for Error {
     fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
         match self {
-            Error::FileTooLarge => None,
-            Error::FileNotInCache => None,
             Error::Io(ref e) => Some(e),
+            _ => None,
```
Comment (froydnj) [inline lru-disk-cache/src/lib.rs]: Please revert this change.

File: src/mock_command.rs [hunk 16]
```diff
@@ -517,7 +517,7 @@ impl MockCommandCreator {
 impl CommandCreator for MockCommandCreator {
     type Cmd = MockCommand;
 
-    fn new(_client: &Client) -> MockCommandCreator {
+    fn new(client: &Client) -> MockCommandCreator {
```
Comment (froydnj) [inline src/mock_command.rs]: Surely this argument is still unused?  Why unprefix it?

File: src/server.rs [hunk 17]
```diff
@@ -823,68 +855,169 @@ where
...
+        let path1 = path.clone();
+        let env = env
+            .into_iter()
+            .cloned()
+            .collect::<Vec<(OsString, OsString)>>();
+
+        let resolve_w_proxy = {
+            let compiler_proxies_borrow = self.compiler_proxies.borrow();
+
+            if let Some((compiler_proxy, filetime)) = compiler_proxies_borrow.get(&path) {
```
Comment (froydnj) [inline src/server.rs]: `filetime` is apparently unused and so should be `_`-prefixed.

File: src/compiler/rust.rs [hunk 18]
```diff
@@ -485,6 +494,216 @@ where
...
+where
+    T: CommandCreatorSync,
+{
+    fn resolve_proxied_executable(
+        &self,
+        mut creator: T,
+        cwd: PathBuf,
+        env: &[(OsString, OsString)],
+    ) -> SFuture<Option<(PathBuf, FileTime)>> {
+        // overwrite the PWD env var with the wanted cwd
```
[the remark this answers is not in the document]
Comment (drahnr) [reply src/compiler/rust.rs]: The `PWD` defines the current dir when a process is launched, as such this works (tested).

File: src/compiler/compiler.rs [hunk 19]
```diff
@@ -1065,13 +1154,15 @@ mod test {
         let f = TestFixture::new();
         let creator = new_creator();
         let pool = CpuPool::new(1);
+        let cwd: PathBuf = "/tmp".into();
```
[the remark this answers is not in the document]
Comment (drahnr) [reply src/compiler/compiler.rs]: API changed and this is a remainder in the tests.

**Comment by drahnr** [comment]
> Please don't mix in unrelated changes (e.g. the `Error::description` changes, the `Error::cause` changes, etc.); it's much much easier to just handle those as a small, separate PR.

In order for CI to pass for `+beta`, I had to merge in the error changes (see comment https://github.com/mozilla/sccache/pull/666#issuecomment-597812065 )

**Comment by froydnj** [comment]
> > Please don't mix in unrelated changes (e.g. the `Error::description` changes, the `Error::cause` changes, etc.); it's much much easier to just handle those as a small, separate PR.
> 
> In order for CI to pass for `+beta`, I had to merge in the error changes (see [#666 (comment)](https://github.com/mozilla/sccache/pull/666#issuecomment-597812065) )

Please submit a new PR and rebase, then.

File: src/compiler/rust.rs (hunk 18 above)
[the remark this answers is not in the document]
Comment (froydnj) [reply src/compiler/rust.rs]: Where is this behavior defined?  I don't see any evidence of this in the Rust standard library, and `env PWD=$HOME ls` (Mac and Linux) doesn't actually list the contents of my home directory.  `environ(7)` says that `PWD` is set by some shells, not that it's some sort of standard interface for changing the directory when exec'ing.

[the remark this answers is not in the document]
Comment (drahnr) [reply src/compiler/rust.rs]: I'll adjust the impl, I apparently had this wrong in the back of my head.

File: src/compiler/compiler.rs [hunk 20]
```diff
@@ -127,6 +128,23 @@ impl<T: CommandCreatorSync> Clone for Box<dyn Compiler<T>> {
     }
 }
 
+pub trait CompilerProxy<T>: Send + 'static
+where
+    T: CommandCreatorSync + Sized,
+{
+    /// maps the executable to be used in `cwd` to the true, proxied compiler
+    ///
+    /// returns the absolute path to the true compiler
```
[the remark this answers is not in the document]
Comment (drahnr) [reply src/compiler/compiler.rs]: I converted this to an error, and expanded the comment for the negative case slightly.

File: src/compiler/compiler.rs (hunk 19 above)
[the remark this answers is not in the document]
Comment (drahnr) [reply src/compiler/compiler.rs]: Removed.

**Comment by drahnr** [comment]
Rebased, removed obsolete changes and resolved all review comments.
